### ISSUE TITLE
Cannot pull image from registries except docker hub.

### ISSUE THREAD

**Issue by jerry-yuan** [containrrr/watchtower#1491]
### Describe the bug

I tried to enable auto upgrade on umami, an opensource web statistic tool, whose image name is `docker.umami.dev/umami-software/umami:mysql-latest`. It's obvious that the image is on `docker.umami.dev`, but watchtower tried to pull image from docker hub and failed.

### Steps to reproduce

1. start a container from `docker.umami.dev/umami-software/umami:mysql-latest`
2. enable auto upgrade on the new container
3. wait trigger auto update

### Expected behavior

pull image from docker.umami.dev

### Environment

- Platform: Ubuntu 20.04
- Architecture: amd64
- Docker Version: 20.10.21

### Your logs

```text
time="2022-12-04T16:38:53+08:00" level=warning msg="Using an HTTP url for Gotify is insecure"
time="2022-12-04T16:38:53+08:00" level=debug msg="Sleeping for a second to ensure the docker api client has been properly initialized."
time="2022-12-04T16:38:54+08:00" level=debug msg="Making sure everything is sane before starting"
time="2022-12-04T16:38:54+08:00" level=debug msg="Retrieving running containers"
time="2022-12-04T16:38:54+08:00" level=info msg="Watchtower 1.5.1"
time="2022-12-04T16:38:54+08:00" level=info msg="Using notifications: gotify"
time="2022-12-04T16:38:54+08:00" level=info msg="Only checking containers using enable label"
time="2022-12-04T16:38:54+08:00" level=info msg="Running a one time update."
time="2022-12-04T16:38:54+08:00" level=debug msg="Checking containers for updated images"
time="2022-12-04T16:38:54+08:00" level=debug msg="Retrieving running containers"
time="2022-12-04T16:38:54+08:00" level=debug msg="Trying to load authentication credentials." container=/phpMyAdmin image="docker.io/phpmyadmin:latest"
time="2022-12-04T16:38:54+08:00" level=debug msg="No credentials for docker.io found" config_file=/config.json
time="2022-12-04T16:38:54+08:00" level=debug msg="Got image name: docker.io/phpmyadmin:latest"
time="2022-12-04T16:38:54+08:00" level=debug msg="Checking if pull is needed" container=/phpMyAdmin image="docker.io/phpmyadmin:latest"
time="2022-12-04T16:38:54+08:00" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"
time="2022-12-04T16:38:55+08:00" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2022-12-04T16:38:55+08:00" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2022-12-04T16:38:55+08:00" level=debug msg="Setting scope for auth token" image=docker.io/phpmyadmin scope="repository:library/phpmyadmin:pull"
time="2022-12-04T16:38:55+08:00" level=debug msg="No credentials found."
time="2022-12-04T16:38:56+08:00" level=debug msg="Parsing image ref" host=index.docker.io image=docker.io/phpmyadmin normalized="docker.io/library/phpmyadmin:latest" tag=latest
time="2022-12-04T16:38:56+08:00" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/library/phpmyadmin/manifests/latest"
time="2022-12-04T16:38:56+08:00" level=debug msg="Found a remote digest to compare with" remote="sha256:3792514e6f6d38819dd8fbb659386bd8ef0019cc7e3fb44ea7c165ff0271ee4f"
time="2022-12-04T16:38:56+08:00" level=debug msg=Comparing local="sha256:3792514e6f6d38819dd8fbb659386bd8ef0019cc7e3fb44ea7c165ff0271ee4f" remote="sha256:3792514e6f6d38819dd8fbb659386bd8ef0019cc7e3fb44ea7c165ff0271ee4f"
time="2022-12-04T16:38:56+08:00" level=debug msg="Found a match"
time="2022-12-04T16:38:56+08:00" level=debug msg="No pull needed. Skipping image."
time="2022-12-04T16:38:56+08:00" level=debug msg="No new images found for /phpMyAdmin"
time="2022-12-04T16:38:56+08:00" level=debug msg="Trying to load authentication credentials." container=/Umami image="docker.io/docker.umami.dev/umami-software/umami:mysql-latest"
time="2022-12-04T16:38:56+08:00" level=debug msg="No credentials for docker.io found" config_file=/config.json
time="2022-12-04T16:38:56+08:00" level=debug msg="Got image name: docker.io/docker.umami.dev/umami-software/umami:mysql-latest"
time="2022-12-04T16:38:56+08:00" level=debug msg="Checking if pull is needed" container=/Umami image="docker.io/docker.umami.dev/umami-software/umami:mysql-latest"
time="2022-12-04T16:38:56+08:00" level=debug msg="Building challenge URL" URL="https://index.docker.io/v2/"
time="2022-12-04T16:38:57+08:00" level=debug msg="Got response to challenge request" header="Bearer realm=\"https://auth.docker.io/token\",service=\"registry.docker.io\"" status="401 Unauthorized"
time="2022-12-04T16:38:57+08:00" level=debug msg="Checking challenge header content" realm="https://auth.docker.io/token" service=registry.docker.io
time="2022-12-04T16:38:57+08:00" level=debug msg="Setting scope for auth token" image=docker.io/docker.umami.dev/umami-software/umami scope="repository:docker.umami.dev/umami-software/umami:pull"
time="2022-12-04T16:38:57+08:00" level=debug msg="No credentials found."
time="2022-12-04T16:38:57+08:00" level=debug msg="Parsing image ref" host=index.docker.io image=docker.io/docker.umami.dev/umami-software/umami normalized="docker.io/docker.umami.dev/umami-software/umami:mysql-latest" tag=mysql-latest
time="2022-12-04T16:38:57+08:00" level=debug msg="Doing a HEAD request to fetch a digest" url="https://index.docker.io/v2/docker.umami.dev/umami-software/umami/manifests/mysql-latest"
time="2022-12-04T16:38:58+08:00" level=warning msg="Could not do a head request for \"docker.io/docker.umami.dev/umami-software/umami:mysql-latest\", falling back to regular pull." container=/Umami image="docker.io/docker.umami.dev/umami-software/umami:mysql-latest"
time="2022-12-04T16:38:58+08:00" level=warning msg="Reason: registry responded to head request with \"401 Unauthorized\", auth: \"Bearer realm=\\\"https://auth.docker.io/token\\\",service=\\\"registry.docker.io\\\",scope=\\\"repository:docker.umami.dev/umami-software/umami:pull\\\",error=\\\"insufficient_scope\\\"\"" container=/Umami image="docker.io/docker.umami.dev/umami-software/umami:mysql-latest"
time="2022-12-04T16:38:58+08:00" level=debug msg="Pulling image" container=/Umami image="docker.io/docker.umami.dev/umami-software/umami:mysql-latest"
time="2022-12-04T16:39:02+08:00" level=debug msg="No new images found for /Umami"
time="2022-12-04T16:39:02+08:00" level=info msg="Session done" Failed=0 Scanned=2 Updated=0 notify=no
time="2022-12-04T16:39:02+08:00" level=info msg="Waiting for the notification goroutine to finish" notify=no
```

**Comment by piksel** [comment]
The error is for the HEAD request which, as you pointed out doesn't understand that your prefix is a docker repo. If that fails, it falls back on regular pulls using the docker daemon and should resolve the image the same way as `docker run` etc. would. So, watchtower should still be working normally, except for that warning.

I don't know how repositories are identified currently, but watchtowers interpretation should align with the docker one.

**Comment by jerry-yuan** [comment]
> The error is for the HEAD request which, as you pointed out doesn't understand that your prefix is a docker repo. If that fails, it falls back on regular pulls using the docker daemon and should resolve the image the same way as `docker run` etc. would. So, watchtower should still be working normally, except for that warning.
> 
> I don't know how repositories are identified currently, but watchtowers interpretation should align with the docker one.

Thanks for your reply. I think there may be no issue in watchtower's image name resolving.

I have another docker management service named portainer. In view of Portainer's behavior, every images pulled from outer registries by `docker pull` or `docker run` has a tag with prefix `docker.io`, while those images build on local machine or pulled by third-party image downloaders like `graboid` and load by `docker load`  does not (the images in second picture with red rectangle).
![Snipaste_2022-12-06_11-07-21](https://user-images.githubusercontent.com/13380647/205802722-a12cf23f-d9de-443e-84da-4f9df2b4233f.png)
![image](https://user-images.githubusercontent.com/13380647/205803024-11f17add-e84b-47a4-9b9b-2208e3b8bb7b.png)

May be the issue is in the compatiblity between third-party docker client and docker daemon?

**Comment by piksel** [comment]
Some other tool seem to have added the `docker.io/` prefix to the container configuration, which is what both Portainer and watchtower uses.

The normalization that adds the prefix to any images without a explicit repository is the exact same code between watchtower and docker (we are importing their library).
I made a playground with some examples of how it transforms it:
https://goplay.tools/snippet/HPrZ08zQ27r

Given that, I suspect that your third party tool normalizes the image names using another method.

**Comment by jerry-yuan** [comment]
@piksel I found another issues like this but the image that cannot pull is from docker.io.
First of all, I have tried to pull the image by `docker pull` command, it said that it has find the image in registry and local image is up to date:
![Screenshot_2023-01-27-15-32-31-58_84d3000e3f4017145260f7618db1d683.jpg](https://user-images.githubusercontent.com/13380647/215032316-d084d981-3d26-4cc0-a6eb-8332ad076a8a.jpg)

But when watchtower try to pull image it produced a 404 not found error.
![Screenshot_2023-01-27-16-29-09-07_f78104b9cb08305e57e6798400aed5f9.jpg](https://user-images.githubusercontent.com/13380647/215042133-2a213320-1bb6-458b-a1e6-0253b929ea58.jpg)

Even though it will fall back to regular pull, a notify per hour is quite annoying.😂 Is there any possible to fix this issue?
